### PR TITLE
Treat dropped chunks as not found

### PR DESCRIPTION
**Pull request by mkindahl**
This commit fixes chunk_scan_find to not return chunks that are
marked as dropped, since the callers don't expect such chunks to be returned as found.

### CODE REVIEW

**Comment by mkindahl** [comment]
This is intended to be a more defensive measure to avoid trying to compress dropped chunks, but it is hard, if not impossible, to create a good test for this.

**Comment by mkindahl** [comment]
> Since compress_chunk/decompress_chunk require valid oids this cannot happen.

I am not so sure.

I can imagine a race condition where a valid OID is passed to the function but while executing, but a concurrent `drop_chunks` execute to completion and mark the chunk as dropped in the metadata and remove the actual chunk from `pg_class`. I do agree it is hard to reproduce though, and I am not entirely sure what kind of transactional guarantees are offered for functions that accept a `REGCLASS` parameter. 

> I think the NULL checks should remain.

Why? The function is strict so the parameter values will never be null.

File: tsl/test/expected/compression_bgw.out [hunk 1]
```diff
@@ -237,6 +237,7 @@ FROM conditions
 GROUP BY device, time_bucket(INTERVAL '1 hour', "time");
 NOTICE:  adding index _materialized_hypertable_8_device_day_idx ON _timescaledb_internal._materialized_hypertable_8 USING BTREE(device, day)
 REFRESH MATERIALIZED VIEW conditions_summary;
+WARNING:  REFRESH did not materialize the entire range since it was limited by the max_interval_per_job setting
```
Comment (k-rus) [inline tsl/test/expected/compression_bgw.out]: Why does this PR introduces this warning?

Comment (mkindahl) [reply tsl/test/expected/compression_bgw.out]: Probably because the range was exceeded. The warning is harmless, but annoying. I'll see if I can get rid of it easily.

Comment (k-rus) [reply tsl/test/expected/compression_bgw.out]: This warning wasn't in the output before and I don't anything in the change, which would provoke it.

Comment (erimatnor) [reply tsl/test/expected/compression_bgw.out]: I am also suspicious of this change.

**Comment by mkindahl** [comment]
> > Since compress_chunk/decompress_chunk require valid oids this cannot happen.
> 
> I am not so sure.
> 
> I can imagine a race condition where a valid OID is passed to the function but while executing, but a concurrent `drop_chunks` execute to completion and mark the chunk as dropped in the metadata and remove the actual chunk from `pg_class`. I do agree it is hard to reproduce though, and I am not entirely sure what kind of transactional guarantees are offered for functions that accept a `REGCLASS` parameter.

Just to test, I added a sleep of 10 seconds between starting the function and reading the chunk metadata data as well as between this and executing the actual scan-and-compress.

* If I update the record in the metadata between the function start end fetch of metadata `ts_chunk_get_by_relid` that will be seen in `compress_chunk_wrapper`.
* If I revert this patch and don't drop the chunk between the `ts_chunk_get_by_relid` and `compress_chunk_impl` it will still proceed with dropping the chunk.

This means that if another thread is in the process of marking the chunk as dropped and then subsequently expect to be able to drop it, it will fail to find the chunk and generate an error.

> 
> > I think the NULL checks should remain.
> 
> Why? The function is strict so the parameter values will never be null.

File: tsl/src/compression/compress_utils.c [hunk 2]
```diff
@@ -320,8 +320,12 @@ tsl_compress_chunk_wrapper(Oid chunk_relid, bool if_not_compressed)
 Datum
 tsl_compress_chunk(PG_FUNCTION_ARGS)
 {
-	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Oid uncompressed_chunk_id = PG_GETARG_OID(0);
```
Comment (erimatnor) [inline tsl/src/compression/compress_utils.c]: I don't think the previous code was worse and I see no value in removing the NULL-check. In contrast, the code is no longer protected against changes in the `STRICT` property in the function declaration. Also, how is this related to other changes?

If you want to assert, you can still `Assert(OidIsValid(uncompressed_chunk_id))` below.

File: tsl/test/expected/compression_bgw.out [hunk 3]
```diff
@@ -260,6 +261,7 @@ SELECT COUNT(*) AS dropped_chunks_count
                    14
 (1 row)
 
+-- Test that adding a compress chunks policy
```
Comment (erimatnor) [inline tsl/test/expected/compression_bgw.out]: Incomplete sentence? Not sure why this is added

File: tsl/src/compression/compress_utils.c [hunk 4]
```diff
@@ -330,12 +334,13 @@ tsl_compress_chunk(PG_FUNCTION_ARGS)
...
 tsl_decompress_chunk(PG_FUNCTION_ARGS)
 {
-	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Oid uncompressed_chunk_id = PG_GETARG_OID(0);
 	bool if_compressed = PG_ARGISNULL(1) ? false : PG_GETARG_BOOL(1);
 	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, 0, true);
-	if (NULL == uncompressed_chunk)
-		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
 
+	/* SQL function `decompress_chunk` is declared as STRICT, so
```
Comment (erimatnor) [inline tsl/src/compression/compress_utils.c]: I would elide this comment. We have many `STRICT` functions and we don't generally have comments like this for other functions. The assertion conveys enough information IMO, and also protects against changes in `STRICT` that would make this comment outdated.

Comment (mkindahl) [reply tsl/src/compression/compress_utils.c]: Removed.

File: tsl/test/expected/compression_bgw.out (hunk 3 above)
Comment (mkindahl) [reply tsl/test/expected/compression_bgw.out]: Removed.

File: tsl/src/compression/compress_utils.c (hunk 2 above)
Comment (mkindahl) [reply tsl/src/compression/compress_utils.c]: The function is declared strict so I see little value in having additional checks and treating `InvalidOid` as an acceptable value for further processing, it just suggest that the function can handle NULL values. It is also not was PostgreSQL itself is doing. However, I removed this from the commit since is not important for the main issue of this commit.

Comment (mkindahl) [reply tsl/src/compression/compress_utils.c]: The function is declared strict so I see little value in having additional checks and treating `InvalidOid` as an acceptable value for further processing, it just suggest that the function can handle NULL values. It is also not was PostgreSQL itself is doing. However, I removed this from the commit since is not important for the main issue of this commit.

File: tsl/test/expected/compression_bgw.out (hunk 1 above)
Comment (mkindahl) [reply tsl/test/expected/compression_bgw.out]: It disappeared when I removed the changes to `compress_utils`. I think this is a test error but will not fix it as part of this commit.

**Comment by mkindahl** [comment]
> Great that you updated the commit message and title. For future reference, can you also apply this update to the PR (this isn't automatically updated)?

Use the message that @k-rus suggested and updated the PR accordingly.

**Comment by mkindahl** [comment]
> Agree with Erik's comment that there might be other place that should get the same treatment.

Yup, but we'll take that in another PR.
